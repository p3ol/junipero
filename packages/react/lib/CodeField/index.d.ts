import { ReactNode, MutableRefObject, ComponentPropsWithRef } from 'react';

export declare type CodeFieldRef = {
  dirty: boolean;
  isJunipero: boolean;
  valid: boolean;
  value: string;
  focus(index: number): void;
  blur(index: number): void;
  reset(): void;
  innerRef: MutableRefObject<any>;
  inputsRef: Array<MutableRefObject<any>>;
};

declare interface CodeFieldProps extends ComponentPropsWithRef<any> {
  autoFocus?: boolean;
  disabled?: boolean;
  className?: string;
  id?: string;
  name?: string;
  required?: boolean;
  size?: number;
  valid?: boolean;
  value?: string;
  onValidate?(
    value: string,
    { dirty, required }: { dirty?: boolean; required?: boolean }
  ): boolean;
  onChange?(changeProps: { value?: string; valid: boolean }): void;
  onPaste?(e: Event): void;
  onFocus?(e: Event): void;
  onBlur?(e: Event): void;
  ref?: MutableRefObject<CodeFieldRef | undefined>;
}

declare function CodeField(props: CodeFieldProps): ReactNode | JSX.Element;

export default CodeField;
