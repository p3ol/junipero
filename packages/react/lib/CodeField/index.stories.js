import { action } from '@storybook/addon-actions';

import CodeField from './index';

export default { title: 'junipero/CodeField' };

export const basic = () => (
  <CodeField onChange={action('change')} />
);

export const autoFocused = () => (
  <CodeField autoFocus onChange={action('change')} />
);

export const withValue = () => (
  <CodeField value="253453" onChange={action('change')} />
);

export const disabled = () => (
  <CodeField disabled value="253453" />
);

export const withValidation = () => (
  <CodeField validate={val => /^[0-9]+$/g.test(val)} />
);
