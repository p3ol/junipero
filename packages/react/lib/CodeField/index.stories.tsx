import { action } from 'storybook/actions';

import Label from '../Label';
import Abstract from '../Abstract';
import FieldControl from '../FieldControl';
import CodeField from './index';

export default { title: 'react/CodeField' };

export const basic = () => (
  <CodeField onChange={action('change')} />
);

export const autoFocused = () => (
  <CodeField autoFocus onChange={action('change')} />
);

export const withValue = () => (
  <CodeField value="253453" onChange={action('change')} />
);

export const disabled = () => (
  <CodeField disabled value="253453" />
);

export const withValidation = () => (
  <CodeField onValidate={val => /^[0-9]+$/g.test(val)} />
);

export const withLabelAndAbstract = () => (
  <FieldControl>
    <Label className="info" htmlFor="code">One-time code</Label>
    <CodeField
      id="code"
      name="code"
      onChange={action('change')}
      required
    />
    <Abstract className="info">
      Here is a little help writing your name
    </Abstract>
  </FieldControl>
);
