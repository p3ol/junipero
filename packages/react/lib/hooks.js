import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
