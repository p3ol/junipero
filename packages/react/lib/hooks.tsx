import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
