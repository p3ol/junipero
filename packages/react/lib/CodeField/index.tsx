import {
  forwardRef,
  useReducer,
  useEffect,
  useMemo,
  useRef,
  useImperativeHandle,
  MutableRefObject,
  ComponentPropsWithRef,
} from 'react';
import { classNames, mockState } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';
import { ForwardedProps } from '../utils';

export declare type CodeFieldRef = {
  dirty: boolean;
  isJunipero: boolean;
  valid: boolean;
  value: string;
  focus(index: number): void;
  blur(index: number): void;
  reset(): void;
  innerRef: MutableRefObject<any>;
  inputsRef: MutableRefObject<Array<any>>;
};

declare interface CodeFieldProps extends ComponentPropsWithRef<any> {
  autoFocus?: boolean;
  disabled?: boolean;
  className?: string;
  id?: string;
  name?: string;
  required?: boolean;
  size?: number;
  valid?: boolean;
  value?: string;
  onValidate?(
    value: string,
    { dirty, required }: { dirty?: boolean; required?: boolean }
  ): boolean;
  onChange?(changeProps: { value?: string; valid: boolean }): void;
  onPaste?(e: Event): void;
  onFocus?(e: Event): void;
  onBlur?(e: Event): void;
  ref?: MutableRefObject<CodeFieldRef | undefined>;
}

const CodeField = forwardRef(({
  className,
  id,
  name,
  value,
  valid,
  autoFocus = false,
  disabled = false,
  required = false,
  size = 6,
  onValidate = (val, { required } = {}) => !!val || !required,
  onChange,
  onPaste,
  onFocus,
  onBlur,
  ...rest
}: CodeFieldProps, ref) => {
  const innerRef = useRef();
  const inputsRef = useRef([]);
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    valid: valid ?? false,
    values: value?.split('').slice(0, size) || [],
    dirty: false,
    active: autoFocus ? 0 : -1,
  });

  useImperativeHandle(ref, () => ({
    value: state.values?.join(''),
    dirty: state.dirty,
    valid: state.valid,
    focus,
    blur,
    reset,
    innerRef,
    inputsRef,
    inputRef,
    isJunipero: true,
  }));

  useEffect(() => {
    if (value) {
      state.values = value.split('').slice(0, size);
      state.valid = onValidate(value, { dirty: state.dirty, required });
      dispatch({ values: state.values, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const focus = (index = 0) => {
    inputsRef.current[index]?.focus();
  };

  const blur = index => {
    inputsRef.current[index]?.blur();
  };

  const reset = () => {
    dispatch({
      dirty: false,
      values: value?.split('').slice(0, size) || [],
      valid: valid ?? false,
    });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const onChange_ = (index, e) => {
    if (disabled) {
      return;
    }

    state.values[index] = e?.target?.value || '';
    state.valid = onValidate?.(state.values.join(''),
      { dirty: state.dirty, required }) || false;
    dispatch({ values: state.values, dirty: true, valid: state.valid });
    onChange?.({ value: state.values.join(''), valid: state.valid });
    updateControl?.({ dirty: true, valid: state.valid });

    if (state.values[index]) {
      focus(index + 1);
    }
  };

  const onKeyDown_ = (index, e) => {
    if (disabled) {
      return;
    }

    const current = inputsRef.current?.[index];
    const prev = inputsRef.current?.[index - 1];
    const next = inputsRef.current?.[index + 1];

    switch (e.key) {
      case 'Backspace':
        if (
          current.selectionStart !== current.selectionEnd ||
          current.selectionStart === 1 ||
          index === 0
        ) {
          return;
        }

        onChange_(index - 1, { target: { value: '' } });
        prev?.focus();
        break;

      case 'ArrowLeft':
        if (current.selectionStart !== current.selectionEnd || index === 0) {
          return;
        }

        prev.selectionStart = current.selectionStart;
        prev.selectionEnd = current.selectionStart;
        prev.focus();
        break;

      case 'ArrowRight':
        if (
          current.selectionStart !== current.selectionEnd ||
          index === size - 1
        ) {
          return;
        }

        next.selectionStart = current.selectionStart;
        next.selectionEnd = current.selectionStart;
        next.focus();
        break;
    }
  };

  const onPaste_ = e => {
    e.preventDefault();
    const value = e.clipboardData.getData('text/plain');

    state.values = value.split('').slice(0, size);
    state.valid = onValidate(value, { dirty: state.dirty, required });
    dispatch({ values: state.values, valid: state.valid });
    focus(Math.min(size - 1, state.values.length));
    onChange?.({ value: state.values.join(''), valid: state.valid });
    onPaste?.(e);
  };

  const onFocus_ = (index, e) => {
    dispatch({ active: index });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (index, e) => {
    dispatch({ active: -1 });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const isEmpty = () =>
    state.values?.length > 0;

  const inputs = useMemo(() => Array.from({ length: size }), [size]);

  return (
    <div
      { ...rest }
      className={classNames(
        'junipero',
        'code-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.active !== -1, disabled },
        className
      )}
      ref={innerRef}
    >
      { inputs.map((_, index) => (
        <input
          ref={ref => { inputsRef.current[index] = ref; }}
          disabled={disabled}
          size={1}
          maxLength={1}
          autoFocus={index === 0 && autoFocus}
          type="tel"
          className={classNames({ active: index === state.active })}
          key={index}
          value={state.values[index] || ''}
          required={required}
          onChange={onChange_.bind(null, index)}
          onKeyDown={onKeyDown_.bind(null, index)}
          onFocus={onFocus_.bind(null, index)}
          onBlur={onBlur_.bind(null, index)}
          onPaste={onPaste_}
        />
      )) }
      <input
        ref={inputRef}
        type="hidden"
        name={name}
        id={id}
        value={state.values?.join?.('') || ''}
      />
    </div>
  );
}) as ForwardedProps<CodeFieldProps, CodeFieldRef>;

CodeField.displayName = 'CodeField';
CodeField.propTypes = {
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  name: PropTypes.string,
  required: PropTypes.bool,
  size: PropTypes.number,
  valid: PropTypes.bool,
  value: PropTypes.string,
  onValidate: PropTypes.func,
  onChange: PropTypes.func,
  onPaste: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
};

export default CodeField;
