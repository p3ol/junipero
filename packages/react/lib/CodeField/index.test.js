import { createRef } from 'react';
import { fireEvent, render } from '@testing-library/react';

import CodeField from './index';

describe('<CodeField />', () => {
  it('should render', () => {
    const { container, unmount } = render(<CodeField />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should correctly fire onChange event', () => {
    const onChange = jest.fn();
    const { container, unmount } = render(<CodeField onChange={onChange} />);

    fireEvent.change(container.querySelectorAll('input')[0],
      { target: { value: '1' } });
    expect(onChange)
      .toHaveBeenCalledWith(expect.objectContaining({ value: '1' }));
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should not fire onChange event if field is disabled', () => {
    const onChange = jest.fn();
    const { container, unmount } = render(
      <CodeField disabled onChange={onChange} />
    );

    fireEvent.change(container.querySelectorAll('input')[0],
      { target: { value: '1' } });
    expect(onChange).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should automatically focus field when autoFocus is enabled', () => {
    const { container, unmount } = render(<CodeField autoFocus={true} />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should automatically switch field focus when changing value', () => {
    const { container, unmount } = render(<CodeField autoFocus={true} />);
    expect(container).toMatchSnapshot();
    fireEvent.change(container.querySelectorAll('input')[0],
      { target: { value: '1' } });
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should allow navigating between inputs with arrow keys', () => {
    const { container, unmount } = render(<CodeField autoFocus={true} />);

    expect(container).toMatchSnapshot();
    fireEvent.keyDown(container.querySelectorAll('input')[0],
      { key: 'ArrowRight' });

    expect(container).toMatchSnapshot();
    fireEvent.keyDown(container.querySelectorAll('input')[1],
      { key: 'ArrowRight' });

    expect(container).toMatchSnapshot();
    fireEvent.keyDown(container.querySelectorAll('input')[2],
      { key: 'ArrowLeft' });

    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should erase chars & move selection when hitting backspace', () => {
    const fieldRef = createRef();
    const { container, unmount } = render(
      <CodeField ref={fieldRef} value="224" />
    );
    expect(container).toMatchSnapshot();

    fireEvent.keyDown(container.querySelectorAll('input')[3],
      { key: 'Backspace' });

    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should not allow to move selection if field is disabled', () => {
    document.activeElement?.blur();
    const { container, unmount } = render(<CodeField disabled />);

    fireEvent.keyDown(container.querySelectorAll('input')[0],
      { key: 'ArrowRight' });

    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should set field as invalid if validation fails', () => {
    const { container, unmount } = render(
      <CodeField onValidate={val => /^[0-9]+$/g.test(val)} />
    );

    fireEvent.change(container.querySelectorAll('input')[0],
      { target: { value: '1' } });

    expect(container).toMatchSnapshot();

    fireEvent.change(container.querySelectorAll('input')[0],
      { target: { value: 'a' } });

    expect(container).toMatchSnapshot();

    unmount();
  });
});
