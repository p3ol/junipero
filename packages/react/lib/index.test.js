import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
