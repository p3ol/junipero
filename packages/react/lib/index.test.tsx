import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
