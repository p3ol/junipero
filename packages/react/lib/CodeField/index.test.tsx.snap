// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CodeField /> > should allow navigating between inputs with arrow keys 1`] = `
<div>
  <div
    class="junipero code-field pristine valid focused"
  >
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should allow navigating between inputs with arrow keys 2`] = `
<div>
  <div
    class="junipero code-field pristine valid focused"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should allow navigating between inputs with arrow keys 3`] = `
<div>
  <div
    class="junipero code-field pristine valid focused"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should allow navigating between inputs with arrow keys 4`] = `
<div>
  <div
    class="junipero code-field pristine valid focused"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should automatically focus field when autoFocus is enabled 1`] = `
<div>
  <div
    class="junipero code-field pristine valid focused"
  >
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should automatically switch field focus when changing value 1`] = `
<div>
  <div
    class="junipero code-field pristine valid focused"
  >
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should automatically switch field focus when changing value 2`] = `
<div>
  <div
    class="junipero code-field dirty valid empty focused"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="1"
    />
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value="1"
    />
  </div>
</div>
`;

exports[`<CodeField /> > should correctly fire onChange event 1`] = `
<div>
  <div
    class="junipero code-field dirty valid empty focused"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="1"
    />
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value="1"
    />
  </div>
</div>
`;

exports[`<CodeField /> > should erase chars & move selection when hitting backspace 1`] = `
<div>
  <div
    class="junipero code-field pristine valid empty"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="2"
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="2"
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="4"
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value="224"
    />
  </div>
</div>
`;

exports[`<CodeField /> > should erase chars & move selection when hitting backspace 2`] = `
<div>
  <div
    class="junipero code-field dirty valid empty focused"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="2"
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="2"
    />
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value="22"
    />
  </div>
</div>
`;

exports[`<CodeField /> > should not allow to move selection if field is disabled 1`] = `
<div>
  <div
    class="junipero code-field pristine valid disabled"
  >
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should not fire onChange event if field is disabled 1`] = `
<div>
  <div
    class="junipero code-field pristine valid disabled"
  >
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      disabled=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should render 1`] = `
<div>
  <div
    class="junipero code-field pristine valid"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<CodeField /> > should set field as invalid if validation fails 1`] = `
<div>
  <div
    class="junipero code-field dirty valid empty focused"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="1"
    />
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value="1"
    />
  </div>
</div>
`;

exports[`<CodeField /> > should set field as invalid if validation fails 2`] = `
<div>
  <div
    class="junipero code-field dirty invalid empty focused"
  >
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value="a"
    />
    <input
      class="active"
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      class=""
      maxlength="1"
      size="1"
      type="tel"
      value=""
    />
    <input
      type="hidden"
      value="a"
    />
  </div>
</div>
`;
